File: data/Pandora.Api/V1/Helpers/ObjectDefinition.cs
using System;
using System.Text.Json.Serialization;
using Pandora.Data.Models;

namespace Pandora.Api.V1.Helpers;

public static class ApiObjectDefinitionMapper
{
    internal static ApiObjectDefinition? Map(ObjectDefinition? input)
    {
        if (input == null)
        {
            return null;
        }

        var definition = new ApiObjectDefinition
        {
            ReferenceName = input.ReferenceName,
            Type = MapApiObjectType(input.Type)
        };
        if (input.NestedItem != null)
        {
            definition.NestedItem = Map(input.NestedItem);
        }
        return definition;
    }

    private static string MapApiObjectType(ObjectType input)
    {
        switch (input)
        {
            case ObjectType.Boolean:
                return ApiObjectType.Boolean.ToString();
            case ObjectType.Csv:
                return ApiObjectType.Csv.ToString();
            case ObjectType.DateTime:
                return ApiObjectType.DateTime.ToString();
            case ObjectType.Dictionary:
                return ApiObjectType.Dictionary.ToString();
            case ObjectType.Float:
                return ApiObjectType.Float.ToString();
            case ObjectType.Integer:
                return ApiObjectType.Integer.ToString();
            case ObjectType.List:
                return ApiObjectType.List.ToString();
            case ObjectType.RawObject:
                return ApiObjectType.RawObject.ToString();
            case ObjectType.Reference:
                return ApiObjectType.Reference.ToString();
            case ObjectType.String:
                return ApiObjectType.String.ToString();

            // Custom Types
            case ObjectType.Location:
                return ApiObjectType.Location.ToString();
            case ObjectType.RawFile:
                return ApiObjectType.RawFile.ToString();
            case ObjectType.SystemAssignedIdentity:
                return ApiObjectType.SystemAssignedIdentity.ToString();
            case ObjectType.LegacySystemAndUserAssignedIdentityList:
                return ApiObjectType.LegacySystemAndUserAssignedIdentityList.ToString();
            case ObjectType.LegacySystemAndUserAssignedIdentityMap:
                return ApiObjectType.LegacySystemAndUserAssignedIdentityMap.ToString();
            case ObjectType.SystemAndUserAssignedIdentityList:
                return ApiObjectType.SystemAndUserAssignedIdentityList.ToString();
            case ObjectType.SystemAndUserAssignedIdentityMap:
                return ApiObjectType.SystemAndUserAssignedIdentityMap.ToString();
            case ObjectType.SystemOrUserAssignedIdentityList:
                return ApiObjectType.SystemOrUserAssignedIdentityList.ToString();
            case ObjectType.SystemOrUserAssignedIdentityMap:
                return ApiObjectType.SystemOrUserAssignedIdentityMap.ToString();
            case ObjectType.UserAssignedIdentityList:
                return ApiObjectType.UserAssignedIdentityList.ToString();
            case ObjectType.UserAssignedIdentityMap:
                return ApiObjectType.UserAssignedIdentityMap.ToString();
            case ObjectType.Tags:
                return ApiObjectType.Tags.ToString();
            case ObjectType.SystemData:
                return ApiObjectType.SystemData.ToString();
        }

        throw new NotSupportedException($"Unsupported ObjectType {input}");
    }
}

public class ApiObjectDefinition
{
    [JsonPropertyName("nestedItem")]
    public ApiObjectDefinition? NestedItem { get; set; }

    [JsonPropertyName("referenceName")]
    public string? ReferenceName { get; set; }

    [JsonPropertyName("type")]
    public string? Type { get; set; }
}

public enum ApiObjectType
{
    Boolean,
    Csv,
    DateTime,
    Dictionary,
    Integer,
    Float,
    List,
    RawFile,
    RawObject,
    Reference,
    String,

    // Custom Types
    Location,
    SystemAssignedIdentity,
    SystemAndUserAssignedIdentityList,
    SystemAndUserAssignedIdentityMap,
    LegacySystemAndUserAssignedIdentityList,
    LegacySystemAndUserAssignedIdentityMap,
    SystemOrUserAssignedIdentityList,
    SystemOrUserAssignedIdentityMap,
    UserAssignedIdentityList,
    UserAssignedIdentityMap,
    Tags,
    SystemData,
}

File: tools/data-api-repository/repository/internal/models/object_definition.go
// Copyright (c) HashiCorp, Inc.
// SPDX-License-Identifier: MPL-2.0

package models

// ObjectDefinition specifies additional information about a specific Object and any associated nested Objects
type ObjectDefinition struct {
	// ObjectDefinitionType defines what kind of ObjectDefinition this is, such as a Reference, String or List
	Type ObjectDefinitionType `json:"type"`

	// Nullable specifies that this type should be unset by sending `null` as the JSON value.
	Nullable bool `json:"nullable"`

	// ReferenceName is the name of the Constant or Model that this is a reference to
	ReferenceName *string `json:"referenceName"`

	// ReferenceNameIsCommonType indicates whether ReferenceName refers to a common type (true) or a type that is local to the service (false)
	ReferenceNameIsCommonType *bool `json:"referenceNameIsCommonType"`

	// NestedItem is a nested ObjectDefinition when Type is a Dictionary or List
	// NOTE: that it's possible to have deeply-nested ObjectDefinitions, e.g. a List of a List of a Dictionary[String (key, fixed as a string) : Integer (value)
	NestedItem *ObjectDefinition `json:"nestedItem,omitempty"`

	// NOTE: these 3 fields were previously located against the Field but instead should be located against the Object Definition
	// as such we'll need to update the Data API to account for this

	// MaxItems specifies the maximum number of items a CSV/Dictionary/List can have, this field is only relevant for the Terraform Schema
	MaxItems *int `json:"maxItems,omitempty"`

	// MinItems specifies the minimum number of items a CSV/Dictionary/List can have, this field is only relevant for the Terraform Schema
	MinItems *int `json:"minItems,omitempty"`

	// DateFormat specifies the format a date field should have, which allows us to generate helper methods (Get and Set) for this date format
	DateFormat *DateFormat `json:"dateFormat,omitempty"`
}

type ObjectDefinitionType string

const (
	// BooleanObjectDefinitionType signifies that this type is a simple Boolean.
	BooleanObjectDefinitionType ObjectDefinitionType = "Boolean"

	// DateTimeObjectDefinitionType signifies that this field contains a DateTime value.
	DateTimeObjectDefinitionType ObjectDefinitionType = "DateTime"

	// IntegerObjectDefinitionType signifies that this field contains an Integer.
	IntegerObjectDefinitionType ObjectDefinitionType = "Integer"

	// FloatObjectDefinitionType signifies that this field contains a Float.
	FloatObjectDefinitionType ObjectDefinitionType = "Float"

	// RawFileObjectDefinitionType signifies that this field contains a byte value e.g. []byte.
	RawFileObjectDefinitionType ObjectDefinitionType = "RawFile"

	// RawObjectObjectDefinitionType signifies that this field contains an interface value e.g. interface{}.
	RawObjectObjectDefinitionType ObjectDefinitionType = "RawObject"

	// ReferenceObjectDefinitionType signifies that this field points to a Constant or a Model.
	ReferenceObjectDefinitionType ObjectDefinitionType = "Reference"

	// StringObjectDefinitionType signifies that this field contains a String.
	StringObjectDefinitionType ObjectDefinitionType = "String"

	// CsvObjectDefinitionType signifies that this field contains a CSV of simple types e.g. String, Integer, Float.
	CsvObjectDefinitionType ObjectDefinitionType = "Csv"

	// DictionaryObjectDefinitionType signifies that this field contains a Dictionary, the
	// Key for a Dictionary is always a String - the Value Type is defined as Nested Object Definition.
	DictionaryObjectDefinitionType ObjectDefinitionType = "Dictionary"

	// ListObjectDefinitionType signifies that this field contains a List, the List's Value Type is defined as a Nested Object Definition.
	ListObjectDefinitionType ObjectDefinitionType = "List"

	EdgeZoneObjectDefinitionType                                ObjectDefinitionType = "EdgeZone"
	LocationObjectDefinitionType                                ObjectDefinitionType = "Location"
	TagsObjectDefinitionType                                    ObjectDefinitionType = "Tags"
	SystemAssignedIdentityObjectDefinitionType                  ObjectDefinitionType = "SystemAssignedIdentity"
	SystemAndUserAssignedIdentityListObjectDefinitionType       ObjectDefinitionType = "SystemAndUserAssignedIdentityList"
	SystemAndUserAssignedIdentityMapObjectDefinitionType        ObjectDefinitionType = "SystemAndUserAssignedIdentityMap"
	LegacySystemAndUserAssignedIdentityListObjectDefinitionType ObjectDefinitionType = "LegacySystemAndUserAssignedIdentityList"
	LegacySystemAndUserAssignedIdentityMapObjectDefinitionType  ObjectDefinitionType = "LegacySystemAndUserAssignedIdentityMap"
	SystemOrUserAssignedIdentityListObjectDefinitionType        ObjectDefinitionType = "SystemOrUserAssignedIdentityList"
	SystemOrUserAssignedIdentityMapObjectDefinitionType         ObjectDefinitionType = "SystemOrUserAssignedIdentityMap"
	UserAssignedIdentityListObjectDefinitionType                ObjectDefinitionType = "UserAssignedIdentityList"
	UserAssignedIdentityMapObjectDefinitionType                 ObjectDefinitionType = "UserAssignedIdentityMap"
	SystemDataObjectDefinitionType                              ObjectDefinitionType = "SystemData"
	ZoneObjectDefinitionType                                    ObjectDefinitionType = "Zone"
	ZonesObjectDefinitionType                                   ObjectDefinitionType = "Zones"
)

type DateFormat string

const (
	RFC3339DateFormat DateFormat = "RFC3339"
	// TODO: others in the future https://github.com/hashicorp/pandora/issues/8 e.g.
	// RFC3339NanoDateFormat DateFormat = "RFC3339Nano"
)
